### PR TITLE
Antispam features

### PR DESCRIPTION
**Pull request by Athozus**
This is WIP. It follows suggestions listed in #102.

Note that it contains many changes, such as :
- Add `warning`, `disabled` and `muted` colors, respectively `#FF8800`, `#332222` and `#CCCCCC`.
- Enlarge boxes views to have enough space for new (un)mark as spam buttons
- New message view layout for the same reason as before
- Add a `list` setting type. I "cheated" a bit on an if statement so I will rework it later, but it works without any problem at the moment (to be sure it is completely modular).
- Add a new `Spam` group to the settings groups, with the new setting `Mute list` (and a toggle button in message view, not visible yet in the screenshot)

It has a dependency with merge request #127 

I will rebase (with some squashes) because it has some commits that are technically independent from the antispam itself.

More details about changes, check commits messages.

screenshot_20231226_174319
screenshot_20231226_190759

### CODE REVIEW

**Comment by Athozus** [comment]
There are some problems with luacheck (I don't really know how to fix it) but the `table.indexof` feature is available since 5.0.0.

**Comment by Niklp09** [comment]
>(I don't really know how to fix it) but the table.indexof feature is available since 5.0.0.

Luacheck doesn't know some functions since they don't belong to the lua specification.
You must add them manually to the `read_globals` table (`table = {fields = {"indexof"}}`).

**Comment by Athozus** [comment]
Thanks. All is ok for the review ?

**Comment by Athozus** [comment]
Until I do some fixes for that you can report, I plan to do the following steps :
- Rebase #127 
- Rebase this PR on master ("remove" #127 commits on this PR worktree)
- Squash luacheck fixes
- Squash mute_list and its sharing
- Spare `check` into add `check` for settings and implement the function for `mute_list` itself
- Rebase

**Comment by S-S-X** [comment]
> Disable beerchat syncing until muting gets fixed

API should be fixed when https://github.com/mt-mods/beerchat/pull/116 is merged.
If you want you could test it with [upgrade-mute-bugs](https://github.com/mt-mods/beerchat/tree/upgrade-mute-bugs) branch.

**Comment by Athozus** [comment]
> API should be fixed when [mt-mods/beerchat#116](https://github.com/mt-mods/beerchat/pull/116) is merged. If you want you could test it with [upgrade-mute-bugs](https://github.com/mt-mods/beerchat/tree/upgrade-mute-bugs) branch.

It successfully works, thanks ! I git force push to remove the latest commit, because it is soon merged, or I let this fix and re-enable sync when it is merged in beerchat master ?

File: util/settings.lua
Comment (BuckarooBanzay) [inline util/settings.lua:78]: what are you planning to do with this function, this looks like a cpu-hog :thinking:

**Review by BuckarooBanzay — APPROVED** [review]
lgtm if this is properly tested

File: util/settings.lua
Comment (Athozus) [reply util/settings.lua:78]: `check` (the end of the part of code) or `sync` ?
- `check` is to be sure all usernames given are existing
- `sync` is to get data from beerchat, but it is temporary disabled waiting for mute fixes merged

Comment (S-S-X) [reply util/settings.lua:78]: I do still think that data `sync` for muting could be a way better as event based specific checks than a static lists and settings sync.

Additionally for beerchat muting to stay sync globally both ways you do not really have to transfer any static data at all and also already has fairly good performance for most parts when used directly.

So making this side also event based would allow for example (if something in this list does not make sense feel free to ask for clarification):
* getting rid of somewhat complicated data transfer between the mods.
* likely better performance without additional optimizations, caching or usage restrictions.
* no loading or syncing of data for players who do not use `/mail`.
* data only fetched when it is actually needed.
* data always up to date, no synchronization required ever.
* no merging of data and no conflict resolutions for merge conflicts.
* does not have to worry about merging possible offline data updates (external systems for example).
* overall simpler compared to data synchronization.
*  automatic handling both public list (other mods) and private list (mail mod), just user interface could use specific code for public/private distinction.

I've not put a lot thought on config sync system so I'd love to hear what problems it can solve and why it would be preferred over on-demand API calls or other event based checks.

Comment (Athozus) [reply util/settings.lua:78]: Non-using `beerchat` disables the `sync`ing script. This is on purpose : it syncs list only when the setting is called. I don't want to have mail storage outside mail : transfer is better than putting muting lists in beerchat. If beerchat is deleted, everything is out.

Comment (S-S-X) [reply util/settings.lua:78]: > If beerchat is deleted, everything is out.

Did I get it right: basically you want that beerchat configrations keep affecting mail even after beerchat has been removed?

Comment (Athozus) [reply util/settings.lua:78]: :+1: Yes, and it saves resources by syncing only when needed.

Comment (S-S-X) [reply util/settings.lua:78]: It actually very likely wastes resources and has worse performance as it syncs what potentially isn't needed but also it is probably simplest way to keep configurations (both wanted and unwanted) after removing source mod.

File: util/settings.lua
Comment (S-S-X) [inline util/settings.lua:102]: This is bad idea as it will likely change to mod storage (because player meta is very limitated).
For this I'd recommend using API as it will then actually keep working after beerchat's internal storage is updated.

Comment (S-S-X) [reply util/settings.lua:102]: Additionally it introduces another inconsistency as it depends on `/mute` command which is not definition of beerchat mute functionality.

Comment (Athozus) [reply util/settings.lua:102]: Okay because when I did it it was still the buggy muting API. I'm gonna make a commit in a second.

Comment (Athozus) [reply util/settings.lua:102]: EDIT : there's not yet a `mute_player` API command in `plugin/mute.lua`

File: util/settings.lua
Comment (Athozus) [reply util/settings.lua:78]: > It actually very likely wastes resources and has worse performance as it syncs what potentially isn't needed but also it is probably simplest way to keep configurations (both wanted and unwanted) after removing source mod.

If you have a better idea tell me, likely beerchat could have a similar `transfer` function ?

File: util/settings.lua
Comment (S-S-X) [reply util/settings.lua:102]: Muting API was not buggy for muting ever. Only for those shortcut API functions you can use to check if someone is muted.

Actual muting API uses event based checks, shortcut functions were not actually used anywhere (never called internally or externally) and were broken but those are only shortcuts for checking status. Muting is event based, not direct API call (because there's no single configuration for muting so you cannot store configuration for muting anywhere).

So, you implement check by adding this:
```lua
if minetest.get_modpath("beerchat") then
	beerchat.register_callback('before_send_pm', function(name, message, target)
		if mail.get_setting(target, "has_muted_"..name) then
			return false, "Optionally tell the player he can't send direct messages"
		end
	end)
end
```

Then use that check by calling (this shortcut was function buggy before but is now fixed):
this should be added where mail is actually sent
```lua
if beerchat.allow_private_message(from, to) then
	do_actually_deliver_mail_message_to_recipient(from, to, body)
end
```

File: util/settings.lua
Comment (S-S-X) [reply util/settings.lua:78]: > > It actually very likely wastes resources and has worse performance as it syncs what potentially isn't needed but also it is probably simplest way to keep configurations (both wanted and unwanted) after removing source mod.
> 
> If you have a better idea tell me, likely beerchat could have a similar `transfer` function ?

See comment in another thread (code comments) https://github.com/mt-mods/mail/pull/128/files#r1451568438

File: util/settings.lua
Comment (S-S-X) [reply util/settings.lua:102]: Also, you want that beerchat configrations keep affecting mail even after beerchat has been removed:
* This can be easily covered by utilizing `beerchat.allow_private_message(from, to)`. Or you can take a risk and keep syncing from player metadata but better only use this data AFTER beerchat is removed.

You want to sync muting information from mail mod towards beerchat:
* Better not do this but instead allow beechat to check if mail mod wants to mute someone instead of pushing your own stuff towards beerchat.
* If you want, beerchat `/list_muted` command can add functionality to list players muted by mail mod. Configuration sync between beerchat and mail mod is not needed for this. Mail mod does not have to implement its own API for this to work (code posted above is everything you need for this on mail mod side).
* If you want, mail mod can list players muted by any mod that utilized beerchat's muting API. Including server specific customizations. For this you can temporarily disable mail mods "before_send_pm" provider and call `beerchat.allow_private_message(from, to)` to get exact results from every mod connected to beerchat (several different ways to mute player).

I might be able to give other suggestions if there's specific functionality you'd want to have? If it is more just desire to sync because syncing sounds useful then we should probably find out what are the pros and cons for complete data sync.

Comment (Athozus) [reply util/settings.lua:102]: Ok for the suggestions (I already read them but didn't have the time to work on).

I can make a commit based on your suggestions, but I'm won't think it would be much better (we still have to get a copy for ourselves).

The idea is not to push to `master` a buggy PR, however my goal is to have #127 and #128 rebased to have a common base before merging new PRs (I don't want to treat a lot of merging conflicts).

Tell me if you're ok for the rebase. I already did a lot of tests, I fixed what I found, so for me it's stable actually.

Comment (S-S-X) [reply util/settings.lua:102]: Comments are based on current PR target branch which currently is master.

Any rebases on development branches are always fine, no issues with that and should be done whenever you feel it helps with development.

But you can always switch that to another dev branch, also rebase on that dev branch if needed and it is actually what I do too when not sure about something, want to compare different somewhat complicated approaches or just want to get easy view to big picture about possible changes.

Comment (Athozus) [reply util/settings.lua:102]: Thanks for the answer.

For me, it is now operational (might be some lost bugs, as any software), but there's not so much issues at the moment.

It is a big pull request (gonna be rebased, with some squashes btw), so I would like other maintainers advice before doing it (such as @BuckarooBanzay )

Comment (BuckarooBanzay) [reply util/settings.lua:102]: > For me, it is now operational (might be some lost bugs, as any software), but there's not so much issues at the moment.

:+1: 

> It is a big pull request (gonna be rebased, with some squashes btw), so I would like other maintainers advice before doing it (such as @BuckarooBanzay )

just squashing sounds fine, i would even do that into the `master` branch IMO (if the code is properly tested)

**Comment by Athozus** [comment]
@S-S-X @BuckarooBanzay 

I followed what I said, following the suggestion from Buckaroo to squash #127 (I haven't understood when I answered, but after reading again the message it was clear).

Here's my working tree. I want to be sure that it's not polluting `master` branch, but I'm quite satisfied of I what I did, so I consider one approval so one of you will be ok.

**Comment by S-S-X** [comment]
Don't really want approve as long as direct player metadata reading or writing is included. There's anyway one approval already...

If you believe it wont cause issues when metadata is obsoleted and/or cleaned up then probably fine, just keep in mind that it is completely private data that is not considered being part of beerchat API and should be refactored soon to allow missing features.

Additionally just a reminder, I didn't read code well enough to be sure but it seems to loop over all players and then read/write their metadata. Is there a checks in place to make sure it wont try to do it for offline players, I think no which might cause crash if sync is called while any of the registered players is offline.

**Pull request merged**
